pytorch_lightning/trainer/data_loading.py

```diff
@@ -136,14 +136,22 @@ def prepare_dataloader(self, dataloader: Any, shuffle: bool, mode: Optional[Runn
         if isinstance(dataloader, CombinedLoader):
             # apply `prepare_dataloader` on all the collection of loaders
             dataloader.loaders = apply_to_collection(
-                dataloader.loaders, DataLoader, self.prepare_dataloader, shuffle, mode=mode
+                dataloader.loaders, (DataLoader, CycleIterator), self.prepare_dataloader, shuffle, mode=mode
             )
+            # the length need to recomputed across all dataloaders in case of special behavior.
+            dataloader._apply_cycle_iterator_length()
```

Comment:
it seems that we cannot do this length computation at the time of instantiation, of `CombinedLoader` (i.e. in `CombinedLoader._wrap_loaders_max_size_cycle`. 
The result is that the dataloader preparation code is interleaving sections corresponding to different (optional) features and this is probably a sign that we will revisit this code (and problems) the future 😅

Reply:
Yes, improved it slightly, but I agree, might be worth to clean this up.